### PR TITLE
Refactor - `LoadedProgramType::Loaded` and `LoadedProgramOwner`

### PR DESCRIPTION
**Pull request by Lichtso**
#### Problems

1. The loaded program tests currently only test themselves (with `LoadedProgramType::TestLoaded`) instead of the production code.
2. The account owner / loader is monomorphized into `LegacyV0`, `LegacyV1` and `Typed`, which has to be undone everywhere the carried executable is needed. It should be a separate field of `LoadedProgram` instead and that could also be used by other `LoadedProgramType`s.

#### Summary of Changes
- Unifies `LoadedProgramType`s `LegacyV0`, `LegacyV1`, `Typed` and `TestLoaded` into `Loaded`.
- Adds `LoadedProgram::account_owner()` and `LoadedProgramOwner`. Using the pubkey of the loader directly was also an option, but I didn't see any reason to waste storing and comparing 256 bits when there can be at most `log_2(5)` ~ 2.3219 bits of entropy.

### CODE REVIEW

**Comment by dmakarov** [comment]
> The account owner / loader is monomorphized into LegacyV0, LegacyV1 and Typed, which has to be undone everywhere the carried executable is needed. 

Could you, please, explain in plain English what this means?

**Comment by Lichtso** [comment]
What I meant is that instead of a generic `LoadedProgramType::Loaded` which would be paried with `LoadedProgramOwner` like so:
- `program_type: LoadedProgramType::Loaded(executable), account_owner: LoadedProgramOwner::LoaderV1`
- `program_type: LoadedProgramType::Loaded(executable), account_owner: LoadedProgramOwner::LoaderV2`
- `program_type: LoadedProgramType::Loaded(executable), account_owner: LoadedProgramOwner::LoaderV3`
- `program_type: LoadedProgramType::Loaded(executable), account_owner: LoadedProgramOwner::LoaderV4`

we instead have a combined one (basically a form of monomorphization):
- `program_type: LoadedProgramType::LegacyV1(executable)`
- `program_type: LoadedProgramType::LegacyV1(executable)`
- `program_type: LoadedProgramType::LegacyV2(executable)`
- `program_type: LoadedProgramType::Typed(executable)`

So whenever you want to get the contained `executable`, you have to explicitly `match program_type` to destructure them:
- `LoadedProgramType::LegacyV1(executable) => executable,`
- `LoadedProgramType::LegacyV2(executable) => executable,`
- `LoadedProgramType::Typed(executable) => executable,`

File: program-runtime/src/loaded_programs.rs [hunk 1]
```diff
@@ -76,14 +87,8 @@ pub enum LoadedProgramType {
...
-    /// Verified and compiled program of loader-v1 or loader-v2
-    LegacyV0(Executable<InvokeContext<'static>>),
-    /// Verified and compiled program of loader-v3 (aka upgradable loader)
-    LegacyV1(Executable<InvokeContext<'static>>),
-    /// Verified and compiled program of loader-v4
-    Typed(Executable<InvokeContext<'static>>),
-    #[cfg(test)]
-    TestLoaded(ProgramRuntimeEnvironment),
+    /// Verified and compiled program
+    Loaded(Executable<InvokeContext<'static>>),
```
Comment (dmakarov) [inline program-runtime/src/loaded_programs.rs:122]: maybe not to address in this PR, but I think the enum name is misnomer.  A LoadedProgramType value can be 'Loaded', 'Unloaded', or something else. It doesn't make sense when a 'loaded program' is in fact 'unloaded'. It also doesn't make sense that for a LoadedProgramType you add an enum value 'Loaded'.  If only 'Loaded' programs are loaded, then what are other 'LoadedProgramType' values?

**Comment by dmakarov** [comment]
> What I meant is that instead of a generic `LoadedProgramType::Loaded` which would be paried with `LoadedProgramOwner` like so:
> 
>     * `program_type: LoadedProgramType::Loaded(executable), account_owner: LoadedProgramOwner::LoaderV1`
> 
>     * `program_type: LoadedProgramType::Loaded(executable), account_owner: LoadedProgramOwner::LoaderV2`
> 
>     * `program_type: LoadedProgramType::Loaded(executable), account_owner: LoadedProgramOwner::LoaderV3`
> 
>     * `program_type: LoadedProgramType::Loaded(executable), account_owner: LoadedProgramOwner::LoaderV4`
> 
> 
> we instead have a combined one (basically a form of monomorphization):
> 
>     * `program_type: LoadedProgramType::LegacyV1(executable)`
> 
>     * `program_type: LoadedProgramType::LegacyV1(executable)`
> 
>     * `program_type: LoadedProgramType::LegacyV2(executable)`
> 
>     * `program_type: LoadedProgramType::Typed(executable)`
> 
> 
> So whenever you want to get the contained `executable`, you have to explicitly `match program_type` to destructure them:
> 
>     * `LoadedProgramType::LegacyV1(executable) => executable,`
> 
>     * `LoadedProgramType::LegacyV2(executable) => executable,`
> 
>     * `LoadedProgramType::Typed(executable) => executable,`

Thanks for the explanation. I'm not convinced that monomorphization is applicable nomenclature in this situation. As far as I understand monomorphization is performed by compiler if a trait can be resolved to a concrete type at compile time, when static dispatch is used.

File: program-runtime/src/loaded_programs.rs (hunk 1 above)
Comment (Lichtso) [reply program-runtime/src/loaded_programs.rs:122]: Yes you are right, it evolved over time and we never adjusted the naming.
How about renaming (in a different PR):
- `LoadedProgram` => `ProgramCacheEntry`
- `LoadedProgramType` => `ProgramCacheEntryType`

Comment (dmakarov) [reply program-runtime/src/loaded_programs.rs:122]: Definitely, sounds better. I don't insist that these name changes should be in this PR. It's up to you.

**Comment by Lichtso** [comment]
It might be a generous view of what "monomorphization" includes, but IMO it means instantiating generics by giving them a distinct code path each. And because the enum items have individual code paths it made sense to me.

**Review by pgarg66 — DISMISSED** [review]

File: program-runtime/src/loaded_programs.rs [hunk 2]
```diff
@@ -355,34 +354,33 @@ impl LoadedProgram {
             metrics.jit_compile_us = jit_compile_time.end_as_us();
         }
 
-        let program = if bpf_loader_deprecated::check_id(loader_key) {
-            LoadedProgramType::LegacyV0(executable)
-        } else if bpf_loader::check_id(loader_key) || bpf_loader_upgradeable::check_id(loader_key) {
-            LoadedProgramType::LegacyV1(executable)
+        let account_owner = if bpf_loader_deprecated::check_id(loader_key) {
```
Comment (alessandrod) [inline program-runtime/src/loaded_programs.rs]: nit: this is LoadedProgramOwner::try_from(loader_key)

File: program-runtime/src/loaded_programs.rs [hunk 3]
```diff
@@ -393,6 +391,7 @@ impl LoadedProgram {
         }
         Some(Self {
             program: LoadedProgramType::Unloaded(self.program.get_environment()?.clone()),
+            account_owner: self.account_owner.clone(),
```
Comment (alessandrod) [inline program-runtime/src/loaded_programs.rs]: nit: the owner type should probably be Copy?

File: program-runtime/src/loaded_programs.rs [hunk 4]
```diff
@@ -464,6 +469,16 @@ impl LoadedProgram {
         let decaying_for = std::cmp::min(63, now.saturating_sub(last_access));
         self.tx_usage_counter.load(Ordering::Relaxed) >> decaying_for
     }
+
+    pub fn account_owner(&self) -> Pubkey {
+        match self.account_owner {
```
Comment (alessandrod) [inline program-runtime/src/loaded_programs.rs]: nit: this is Into<Pubkey> for LoadedProgramOwner

File: program-runtime/src/loaded_programs.rs [hunk 5]
```diff
@@ -1061,16 +1074,19 @@ impl<FG: ForkGraph> ProgramCache<FG> {
...
-                        LoadedProgramType::LegacyV0(_) | LoadedProgramType::LegacyV1(_)
-                            if include_program_runtime_v1 =>
-                        {
-                            Some((*id, program.clone()))
-                        }
-                        LoadedProgramType::Typed(_) if include_program_runtime_v2 => {
-                            Some((*id, program.clone()))
+                        LoadedProgramType::Loaded(_) => {
+                            let include =
+                                if let LoadedProgramOwner::LoaderV4 = program.account_owner {
```
Comment (alessandrod) [inline program-runtime/src/loaded_programs.rs]: nit: if you implement Eq for LoadedProgramOwner then I think this can become one
if with LoaderV4 && include_program_runtime_v2 { .. } else if
include_program_runtime_v1 { ... }

**Review by alessandrod — DISMISSED** [review]
Sorry it took so long to review!

I love this, we're testing more, code looks better and logic is simpler. Thanks a lot for doing it.

I only left a few nits, otherwise looks great.

**Review by alessandrod — APPROVED** [review]
